src/pages/scientificServices/pipelines/utils/upload-utils.ts

```diff
@@ -1,6 +1,65 @@
+import pluralize from 'pluralize';
 import { Dispatch, SetStateAction } from 'react';
 import { PipelineInputFileUploadState } from 'src/pages/scientificServices/pipelines/components/inputs/PipelineFileInput';
 
+// Returns a function that handles progress events for file uploads
+// This function updates the upload state with progress percentage and
+// estimated time remaining using a moving average of upload rates
+function createProgressEventListener(
+  inputName: string,
+  inputFile: File,
+  sessionUrl: string,
+  setUploadState: Dispatch<SetStateAction<Record<string, PipelineInputFileUploadState>>>,
+  startTime: number,
+  uploadRateSamples: number[],
+  lastEtaUpdate: number
+) {
+  return (event: ProgressEvent) => {
+    if (event.lengthComputable) {
+      const currentTime = Date.now();
+      const elapsedTime = currentTime - startTime;
+
+      // Calculate the upload rate of the most recent progress event
+      const totalBytesUploaded = event.loaded;
+      const uploadRate = totalBytesUploaded / elapsedTime; // bytes per ms
+      const percent = Math.round((totalBytesUploaded / inputFile.size) * 100);
+
+      // Only update ETA every 2 seconds or on the first update, to avoid choppy ETA updates
```

Comment:
OoC - do we control how often this event gets "triggered"?  Like how many times will this if statement get run before its been 2 seconds?